annotatedcsv/decode.go

```diff
@@ -0,0 +1,463 @@
+			return nil
+		}
+	case reflect.Slice:
+		s := et.Elem()
+		if s != stringType && s != emptyInterfaceType {
+			return fmt.Errorf("cannot decode into pointer to slice of %v", s)
+		}
+		fkind := fieldKindOf(s)
+		for i, col := range columns {
+			colType, ok := columnTypes[col.Type]
+			if !ok {
+				// ignore invalid type and use string
+				colType = stringCol
+			}
+
+			convert, ok := conversions[conv{colType, fkind}]
+			if !ok {
+				return fmt.Errorf("cannot convert from column type %s to %v", col.Type, s)
+			}
+			setters[i] = func(v reflect.Value, colIndex int) error {
+				if err := r.convertColumnValue(v.Index(colIndex), colIndex, convert); err != nil {
+					return err
+				}
+				return nil
+			}
+		}
+		r.decodeRow = func(v reflect.Value, row []string) error {
+			e := v.Elem()
+			c := len(r.colSetters)
+			if e.Cap() < c {
+				e = reflect.MakeSlice(e.Type(), c, c)
+			} else {
+				e = e.Slice(0, c)
+			}
+			for i := 0; i < c; i++ {
+				if err := r.colSetters[i](e, i); err != nil {
+					return err
+				}
+			}
+			v.Elem().Set(e)
+			return nil
+		}
+	}
+
+	r.colSetters = setters
+	r.decodeType = t
+
+	return nil
+}
+
+// convertColumnValue set a value from current row of given column index
+// to a struct or a slice field value
+func (r *Reader) convertColumnValue(v reflect.Value, colIndex int, convert valueSetter) error {
+	if err := convert(v, r.row[colIndex]); err != nil {
+		return fmt.Errorf(`cannot convert value %q in column of type %q to Go type %v at line %d: %w`, r.row[colIndex], r.cols[colIndex].Type, v.Type(), r.r.Line(), err)
+	}
+	return nil
+}
+
+// fieldKind represents one of the possible kinds of struct field.
+// It's similar to reflect.Kind (every reflect.Kind constant i
+// is represented as fieldKind(i)) except that it also has
+// defined constants for field types that have special treatment,
+// such as time.Duration.
+type fieldKind uint
+
+// conv represents a possible conversion source and destination.
+type conv struct {
+	from colType
+	to   fieldKind
+}
+
+// intKinds holds all supported signed integer kinds
+var intKinds = []fieldKind{
+	fieldKind(reflect.Int),
+	fieldKind(reflect.Int8),
+	fieldKind(reflect.Int16),
+	fieldKind(reflect.Int32),
+	fieldKind(reflect.Int64),
+}
+
+// uintKinds holds all supported unsigned integer kinds
+var uintKinds = []fieldKind{
+	fieldKind(reflect.Uint),
+	fieldKind(reflect.Uint8),
+	fieldKind(reflect.Uint16),
+	fieldKind(reflect.Uint32),
+	fieldKind(reflect.Uint64),
+}
+
+// floatKinds holds all supported floating point number kinds
+var floatKinds = []fieldKind{
+	fieldKind(reflect.Float32),
+	fieldKind(reflect.Float64),
+}
+
+// rest of supported type kinds
+const (
+	durationKind = fieldKind(255 + iota)
+	timeKind
+	bytesKind
+)
+
+// colType represents the type of an annotated CSV column.
+type colType uint
+
+const (
+	stringCol = colType(iota)
+	boolCol
+	durationCol
+	longCol
+	unsignedLongCol
+	doubleCol
+	base64BinaryCol
+	rfc3339Col
+	rfc3339NanoCol
```

Comment:
I don't think we need to distinguish between `rfc3339` and `rfc3339Nano` - I'd lose `rfc3330Nano`

Reply:
ok